wit/couchbase/document.wit

```diff
@@ -0,0 +1,347 @@
+    timeout-ns: option<u64>,
+
+    /// How and whether to retry the operation
+    retry-strategy: option<retry-strategy>,
+
+    /// A known span to associate this lookup with
+    parent-span: option<request-span>,
+  }
+
+  /// Replace a document with the given ID with a new document
+  replace: func(
+    id: document-id,
+    doc: document,
+    options: option<document-replace-options>,
+  ) -> result<mutation-metadata, document-error>;
+
+  /////////////////////////
+  /// Document - Upsert ///
+  /////////////////////////
+
+  /// Options for performing a document upsert
+  record document-upsert-options {
+    /// When the document should expire (nanoseconds)
+    expires-in-ns: u64,
+
+    /// Whether to preserve expiration values set on any documents modified by the query
+    preserve-expiry: bool,
+
+    /// (Deprecated)
+    /// Traditional durability settings for persistence, only usable from Couchbase versions earlier than 6.0
+    ///
+    /// See: https://docs.couchbase.com/go-sdk/current/howtos/subdocument-operations.html#1.6@go-sdk::durability.adoc
+    persist-to: u64,
+
+    /// (Deprecated)
+    /// Traditional durability settings for replication, only usable from Couchbase versions earlier than 6.0
+    ///
+    /// See: https://docs.couchbase.com/go-sdk/current/howtos/subdocument-operations.html#1.6@go-sdk::durability.adoc
+    replicate-to: u64,
+
+    /// Durability level for the mutation
+    ///
+    /// (this should be used instead of persist-to/replicate-to)
+    durability-level: durability-level,
+
+    /// Timeout that should be used, in nanoseconds
+    /// Even if a timeout is not specified, the implementer *may* provide a default timeout.
+    timeout-ns: option<u64>,
+
+    /// How and whether to retry the operation
+    retry-strategy: option<retry-strategy>,
+
+    /// A known span to associate this lookup with
+    parent-span: option<request-span>,
+  }
+
+  /// Create or update (replace) an existing document with the given ID
+  upsert: func(
+    id: document-id,
+    doc: document,
+    options: option<document-upsert-options>,
+  ) -> result<mutation-metadata, document-error>;
+
+  //////////////////////
+  /// Document - Get ///
+  //////////////////////
+
+  /// Options for performing a document get
+  record document-get-options {
+    /// Whether expiry should also be retreived
+    with-expiry: bool,
+
+    /// Perform a projection of only specific paths of the result, returning only
+    /// the fields indicated by the given list of paths.
+    project: option<list<string>>,
+
+    /// Timeout that should be used, in nanoseconds
+    /// If not specified, the implementer *may* provide a default timeout.
+    timeout-ns: option<u64>,
+
+    /// How and whether to retry the operation
+    retry-strategy: option<retry-strategy>,
+
+    /// Level of replica read to enable
+    use-replica: option<replica-read-level>,
```

Comment:
I'm having second thoughts about this. SDKs use three separate methods : `get`, `getAnyReplica` and `getAllReplicas`. It is not consistent with the `query` method that uses the `useReplica` option, that's why I suggested this option, but I'm wondering if in order to follow the path of least surprise, we should not also have an interface with three separate methods. WDYT ?

Reply:
Hmnnnn I personally think of a the amount of replicas to require as being an option to a `get` -- so it makes more sense organization wise for me. Was there any particular reason behind the three separate methods? 

If there's nothing that distinguishes them from each other, I think I prefer leaving it as an option to the one method...

Reply:
Well the `getAllReplicas` returns a list with all copies of the document.
So we could have two methods instead of three, but I feel like it's not a good idea to walk away from the known interfaces.

I also noticed that the methods `lookupAnyReplica` and `lookupAllReplicas` are missing.

Reply:
Ah I think I misunderstood -- `getAllReplicas`/`getAnyReplicas` are not just gets... they're more for checking the state/knowing what version of a given object the replicas had/being consistent with that? I guess that's quite different from `get` then.

I think it makes sense to have `use-replica` on `get` as an option *and* those other ones for a different use case.

I'll add lookup functions as well, they mostly fit in the CRUD paradigm that we're going for for v0.1.0

- [ ] Add `getAnyReplicas`, `getAllReplicas`
- [ ] Add `lookupAnyReplica`, `lookupAllReplicas`

Reply:
Hey @JesusTheHun added these separate fns -- let me know what you think!

Reply:
`get` returns the document from the active node or fails if the node is unavailable.
`getAnyReplica` returns a single copy of the document, either from the active node **or** from a replica. 
`getAllReplicas` returns all copies from all nodes which must all be online for this request to succeed.

So the option `use-replica` is redundant with `getAnyReplica`. It's exactly the same thing.